### PR TITLE
feat: extend plugin API supports close widget

### PR DESCRIPTION
**Pull request by airslice**
close https://github.com/reearth/reearth/issues/365

# Overview

Enable plugin to close the main ui widget.

## What I've done

Implement the API:

```ts
reearth.ui.close: () => void;
```

## How I tested

Tested with plugin:

```js
const ui = `
  <style>
    html,
    body {
      margin: 0;
      width: 350px;
      font-size: 14px;
    }
    button {
      margin: 0;
    }
    #wrapper {
      height: 100%;
      box-sizing: border-box;
      border-radius: 12px;
      padding: 12px;
      background: white;
    }
    #wrapper {
      box-sizing: border-box;
      border-radius: 12px;
      background: white;
      height: 100%;
    }
    h3{
      margin: 0 0 5px;
    }
    h4{
      margin: 5px 0;
    }
  </style>

  <div id="wrapper">
    <h3>Close Widget</h3>
    <h4>Resize the viewport and this widget will be visible only when viewport's height >= 500.</h4>
    <h4>Current height is <span id="currentHeight"></span></h4>
  </div>

  <script>
    parent.postMessage({ type: "requestViewport" });

    addEventListener("message", (e) => {
      if (e.source !== parent) return;
      if (e.data.type === "getViewport"){
        document.getElementById("currentHeight").innerHTML = e.data.message.height
      }
    });
  </script>
`;

let visible;

function updateUIState(viewport){
  if(viewport.height >= 500){
    if(visible !== true){
      reearth.ui.show(ui);
      visible = true;
    }
  }else if(visible !== false){
    reearth.ui.close();
    visible = false;
  }
  reearth.ui.postMessage({message:viewport, type: "getViewport"});
}

updateUIState(reearth.viewport);

reearth.on("resize", (viewport) => {
  updateUIState(viewport);
})

reearth.on("message", (msg) => {
  if (msg.type === "requestViewport") {
    reearth.ui.postMessage({message: reearth.viewport, type: "getViewport"});
  }
});
```

### CODE REVIEW

File: src/components/molecules/Visualizer/Plugin/hooks.ts [hunk 1]
```diff
@@ -61,6 +63,8 @@ export default function ({
   const popupVisible = useRef<boolean>(false);
   const externalRef = useRef<HTMLIFrameElement>(null);
 
+  const [uiVisible, onUIShow] = useState<boolean>(!!visible);
```
Comment (KaWaite) [inline src/components/molecules/Visualizer/Plugin/hooks.ts:66]: `onUIShow` sounds like a function that is going to do something when the UI is shown. Is there a better name, like just `setUIVisibility` or `setUI` or something?

File: src/components/atoms/Plugin/PluginIFrame/useIFrame.ts [hunk 2]
```diff
@@ -61,6 +61,10 @@ export default function useIFrame({
     [iframeCanBeVisible, onRender, postMessage],
   );
 
+  useEffect(() => {
+    setIFrameState(state => [state[0], { ...state[1], visible: !!iframeCanBeVisible }]);
```
Comment (KaWaite) [inline src/components/atoms/Plugin/PluginIFrame/useIFrame.ts:65]: It isn't a big deal, but is it okay to name it a bit differently like:
```
setIFrameState(({ html, options }) => [html, { ...options, visible: !!iframeCanBeVisible }]
```

File: src/components/molecules/Visualizer/Plugin/hooks.ts (hunk 1 above)
Comment (airslice) [reply src/components/molecules/Visualizer/Plugin/hooks.ts]: Updated to `setUIVisibility`.

File: src/components/atoms/Plugin/PluginIFrame/useIFrame.ts (hunk 2 above)
Comment (airslice) [reply src/components/atoms/Plugin/PluginIFrame/useIFrame.ts]: Updated.

**Review by KaWaite — APPROVED** [review]

**Review by rot1024 — APPROVED** [review]

**Pull request merged**